File: projects/tox4j/src/test/java/im/tox/gui/events/AddFriendButtonOnAction.java
package im.tox.gui.events;

import im.tox.gui.MainView;
import im.tox.tox4j.core.ToxCoreConstants;
import im.tox.tox4j.core.exceptions.ToxFriendAddException;

import javax.swing.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.util.Arrays;

import static im.tox.tox4j.ToxCoreTestBase.parsePublicKey;

public class AddFriendButtonOnAction implements ActionListener {
  private MainView toxGui;

  public AddFriendButtonOnAction(MainView toxGui) {
    this.toxGui = toxGui;
  }

  @Override
  public void actionPerformed(ActionEvent event) {
    try {
      byte[] publicKey = parsePublicKey(toxGui.friendId.getText());
      int friendNumber;
      if (toxGui.friendRequest.getText().isEmpty()) {
        friendNumber = toxGui.tox.addFriendNoRequest(publicKey);
      } else {
        friendNumber = toxGui.tox.addFriend(publicKey, toxGui.friendRequest.getText().getBytes());
      }
      toxGui.friendListModel.add(friendNumber, Arrays.copyOf(publicKey, ToxCoreConstants.PUBLIC_KEY_SIZE));
      toxGui.addMessage("Added friend number " + friendNumber);
      toxGui.save();
    } catch (ToxFriendAddException e) {
      toxGui.addMessage("Add friend failed: ", e.code());
    } catch (Throwable e) {
      JOptionPane.showMessageDialog(toxGui, MainView.printExn(e));
    }
  }
}

File: projects/tox4j/src/test/java/im/tox/gui/events/AddFriendButtonOnAction.scala
package im.tox.gui.events

import java.awt.event.{ActionEvent, ActionListener}
import javax.swing._

import im.tox.gui.MainView
import im.tox.tox4j.ToxCoreTestBase.parsePublicKey
import im.tox.tox4j.core.ToxCoreConstants
import im.tox.tox4j.core.exceptions.ToxFriendAddException

final class AddFriendButtonOnAction(toxGui: MainView) extends ActionListener {

  override def actionPerformed(event: ActionEvent): Unit = {
    try {
      val publicKey = parsePublicKey(toxGui.friendId.getText)

      val friendNumber =
        if (toxGui.friendRequest.getText.isEmpty) {
          toxGui.tox.addFriendNorequest(publicKey)
        } else {
          toxGui.tox.addFriend(publicKey, toxGui.friendRequest.getText.getBytes)
        }

      toxGui.friendListModel.add(friendNumber, publicKey.slice(0, ToxCoreConstants.PublicKeySize))
      toxGui.addMessage("Added friend number ", friendNumber)
      toxGui.save()
    } catch {
      case e: ToxFriendAddException =>
        toxGui.addMessage("Add friend failed: ", e.code)
      case e: Throwable =>
        JOptionPane.showMessageDialog(toxGui, MainView.printExn(e))
    }
  }

}
